### PR TITLE
Dev pathloss

### PR DESCRIPTION
**Pull request by brunohcfaria**
<h2>Result of the Sprint26</h2>
Now the simulator can import Base Station physical information from an input file as well as the path loss matrix.
The path loss information is implemented as "new" propagation model in the simulation structure.

### CODE REVIEW

File: sharc/propagation/propagation_input_file.py [hunk 1]
```diff
@@ -35,28 +36,76 @@ def __init__(self, input_folder:str):
...
+                    except ValueError as err:
+                        if param in essential_params:
+                            sys.stderr.write(param + 
+                                         " parameter not in path loss file " +
+                                         file + "\n")
+                            sys.exit(1)
+                        else:
+                            idx_list.append(-4)
+                    
+                # Get parameter values
```
Comment (brunohcfaria) [inline sharc/propagation/propagation_input_file.py]: The header parsing is heavily dependent on list indexing. This is very confusing and lacks performance. A better approach should be analysed.

File: sharc/simulation.py [hunk 2]
```diff
@@ -195,15 +196,22 @@ def calculate_coupling_loss(self,
...
+                                                 distance_2D=d_2D,
+                                                 frequency=self.parameters.imt.frequency*np.ones(d_2D.shape),
+                                                 indoor_stations=np.tile(station_b.indoor, (station_a.num_stations, 1)),
+                                                 bs_height=station_a.height,
+                                                 ue_height=station_b.height,
+                                                 elevation=elevation_angles,
+                                                 shadowing=self.parameters.imt.shadowing,
+                                                 line_of_sight_prob=self.parameters.imt.line_of_sight_prob)
+            else:
+                #TODO Pass parameters correctly
```
Comment (CalilQ) [inline sharc/simulation.py]: This TODO can be removed

File: sharc/propagation/propagation_factory.py [hunk 3]
```diff
@@ -41,6 +42,8 @@ def createPropagation(channel_model: str, param: Parameters) -> Propagation:
             return PropagationP619()
         elif channel_model == "INDOOR":
             return PropagationIndoor(param.indoor)
+        elif channel_model == "INPUT_FILES":
```
Comment (CalilQ) [inline sharc/propagation/propagation_factory.py:45]: INPUT_FILE (singular) might be more appropriate

File: sharc/propagation/propagation_input_file.py [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+                                    "LOWER_LEFT",
+                                    "UPPER_RIGHT",
+                                    "RESOLUTION",
+                                    "RECEIVER_GAIN"]
+                idx_list = []
+                for param in param_list:
+                    try:
+                        idx_list.append(head_list.index(param))
+                    except ValueError as err:
+                        if param in essential_params:
```
Comment (brunohcfaria) [inline sharc/propagation/propagation_input_file.py]: Shouldn't be this test outside the except clause?

File: tests/test_propagation_input_file.py [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+import numpy as np
+
+from sharc.propagation.propagation_input_file import PropagationInputFile
+from sharc.support.named_tuples import PathLossHeader
+
+class PropagationInputFileTest(unittest.TestCase):
+    
+    def setUp(self):
+        # Test 1
+        input_folder = "propagation_test_files\\test_1"
```
Comment (brunohcfaria) [inline tests/test_propagation_input_file.py]: Change all file names and directories to OS independent.

File: sharc/propagation/propagation_input_file.py [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+        # Get keywork arguments
+        bs_id = kwargs["bs_id"]
+        ue_position_x = kwargs["ue_position_x"]
+        ue_position_y = kwargs["ue_position_y"]
+
+        # Initialize array
+        loss = np.zeros((len(bs_id),len(ue_position_x)))
+
+        # Loop through all the cells
+        for k in range(len(bs_id)):
```
Comment (brunohcfaria) [inline sharc/propagation/propagation_input_file.py]: Is `bs_id` a list of Base stations ids? If so, use `enumerate(bs_id)` to iterate over bs_id elements and indices.

**Review by CalilQ — APPROVED** [review]
Code style and unit tests are ok. Result consistency and regression still need to be checked though.

File: sharc/propagation/propagation_input_file.py [hunk 7]
```diff
@@ -82,48 +82,37 @@ def __init__(self, input_folder: str):
...
+                head["RESOLUTION"] = float(head["RESOLUTION"][0])
+                head["RECEIVER_GAIN"] = float(head["RECEIVER_GAIN"][0])
                 
                 # Undefined parameter value
                 undefined = np.nan
                 
                 # Parse other parameters
-                try: location = [float(x) for x in head_dict["LOCATION"]]
-                except KeyError: location = undefined
+                try: head["LOCATION"] = [float(x)
```
Comment (brunohcfaria) [inline sharc/propagation/propagation_input_file.py]: @CalilQ O que vc acha de trocar o(s) try/except por isso aqui?
`
if 'LOCATION' not in head.keys():
    head['LOCATION'] = undefined
`
Geralmente a construção try/except é para algo excepcional.

File: sharc/propagation/propagation_input_file.py (hunk 6 above)
Comment (brunohcfaria) [reply sharc/propagation/propagation_input_file.py]: @CalilQ 
em vez de `for k in range(len(ds_id))`, que tal `for k, bs in enumerate(bs_id):`
Vc pode remover a linha 163: `bs = bs_id[k]`
é mais Pythonic...

**Pull request merged**